cranelift/codegen/src/machinst/debug.rs

```diff
@@ -0,0 +1,414 @@
+        trace!("at block {} -> state: {:?}", block, state);
+        let mut iix = label_insn_iix[block as usize];
+        while iix < insts.len() as u32 {
+            state.step(&insts[iix as usize]);
+            trace!(" -> inst #{}: {:?}", iix, insts[iix as usize]);
+            trace!("    --> state: {:?}", state);
+
+            let term = insts[iix as usize].is_term();
+            if term.is_term() {
+                for succ in term.get_succs() {
+                    trace!("    SUCCESSOR block {}", succ.get());
+                    if let Some(succ_state) = block_starts.get_mut(&succ.get()) {
+                        trace!("       orig state: {:?}", succ_state);
+                        if succ_state.intersect_from(&state).changed {
+                            if worklist_set.insert(succ.get()) {
+                                worklist.push_back(succ.get());
+                            }
+                            trace!("        (changed)");
+                        }
+                        trace!("       new state: {:?}", succ_state);
+                    } else {
+                        // First time seeing this block
+                        block_starts.insert(succ.get(), state.clone());
+                        worklist.push_back(succ.get());
+                        worklist_set.insert(succ.get());
+                    }
+                }
+                break;
+            }
+
+            iix += 1;
+        }
+    }
+
+    // Now iterate over blocks one last time, collecting
+    // value-label locations.
+
+    let mut value_labels_ranges: ValueLabelsRanges = HashMap::new();
+    for block in 0..label_insn_iix.len() {
+        let start_iix = label_insn_iix[block];
+        let end_iix = if block == label_insn_iix.len() - 1 {
+            insts.len() as u32
+        } else {
+            label_insn_iix[block + 1]
+        };
+        let block = block as u32;
+        let mut state = block_starts.get(&block).unwrap().clone();
+        for iix in start_iix..end_iix {
+            let offset = inst_start(iix as usize);
+            let end = inst_ends[iix as usize];
+            state.step(&insts[iix as usize]);
+
+            for (label, locs) in &state.label_to_locs {
+                trace!("   inst {} has label {:?} -> locs {:?}", iix, label, locs);
+                // Find an appropriate loc: a register if possible, otherwise pick the first stack
+                // loc.
+                let reg = locs.iter().cloned().find(|l| l.is_reg());
+                let stack = locs.iter().cloned().find(|l| l.is_stack());
+                if let Some(loc) = reg.or(stack) {
+                    let loc = LabelValueLoc::from(loc);
+                    let list = value_labels_ranges.entry(*label).or_insert_with(|| vec![]);
+                    if list.is_empty()
+                        || list.last().unwrap().end <= offset
+                        || list.last().unwrap().loc != loc
```

Comment:
The `map_or(true, ...)` form is a little hard to read for me, so I did `list.last().map(...).unwrap_or(true)` to be a little more explicit, and added a comment too. Happy to make things more concise, just want to avoid Haskell-like terseness that takes too much brain (for me, anyway) to decode :-)